MailCore/Utils/URLSchemeHandler.swift

```diff
@@ -19,43 +19,86 @@
+
+    func didFinish(for url: URL) {
+        urlSchemeTasksInProgress[url.absoluteString]?.didFinish()
+        urlSchemeTasksInProgress[url.absoluteString] = nil
+    }
+
+    func didFailWithError(_ error: Error, for url: URL) {
+        urlSchemeTasksInProgress[url.absoluteString]?.didFailWithError(error)
+        urlSchemeTasksInProgress[url.absoluteString] = nil
+    }
+}
+
 public class URLSchemeHandler: NSObject, WKURLSchemeHandler {
     public static let scheme = "mail-infomaniak"
     public static let domain = "://mail.infomaniak.com"
 
-    private var tasksInProgress = [String: URLSessionDataTask]()
+    private let taskProgressHandler = TaskProgressHandler()
 
     public func webView(_ webView: WKWebView, start urlSchemeTask: WKURLSchemeTask) {
         guard let url = urlSchemeTask.request.url else { return }
-        tasksInProgress[url.absoluteString]?.cancel()
-
-        var components = URLComponents(url: url, resolvingAgainstBaseURL: true)
-        components?.scheme = "https"
-        var request = URLRequest(url: components!.url!)
-        request.addValue(
-            "Bearer \(AccountManager.instance.currentAccount.token.accessToken)",
-            forHTTPHeaderField: "Authorization"
-        )
-        let task = URLSession.shared.dataTask(with: request) { data, response, error in
-            guard (error as? URLError)?.code != .cancelled else { return }
-
-            if let response = response {
-                urlSchemeTask.didReceive(response)
-            }
-            if let data = data {
-                urlSchemeTask.didReceive(data)
-                urlSchemeTask.didFinish()
-            }
-            if let error = error {
-                urlSchemeTask.didFailWithError(error)
+        Task {
+            await taskProgressHandler.cancelDataTask(for: url)
+            await taskProgressHandler.addURLSchemeTask(urlSchemeTask, for: url)
+
+            var components = URLComponents(url: url, resolvingAgainstBaseURL: true)
+            components?.scheme = "https"
+            var request = URLRequest(url: components!.url!)
+            request.addValue(
+                "Bearer \(AccountManager.instance.currentAccount.token.accessToken)",
+                forHTTPHeaderField: "Authorization"
+            )
+            let task = URLSession.shared.dataTask(with: request) { data, response, error in
+                Task { [weak self] in
+                    guard (error as? URLError)?.code != .cancelled else { return }
+
+                    if let response {
+                        await self?.taskProgressHandler.didReceive(response, for: url)
+                    }
+                    if let data {
+                        await self?.taskProgressHandler.didReceive(data, for: url)
+                        await self?.taskProgressHandler.didFinish(for: url)
+                    }
+                    if let error {
+                        await self?.taskProgressHandler.didFailWithError(error, for: url)
```

Comment:
// can we group the two didReceive and didFinish in a dedicated call ?
    await self.taskProgressHandler.didReceiveAndFinish(data, for: url)